calyx-py/calyx/builder.py

```diff
@@ -898,6 +934,48 @@ def mem_latch_d2(self, mem, i, j, groupname):
             latch_grp.done = mem.done
         return latch_grp
 
+    def flatten_idx(self, dims, indices):
+        """Translate an n-dimensional index into a corresponding 1d index"""
+        assert len(dims) == len(indices)
+        i = len(indices) - 1
+        prod = 1
+        total = indices[-1]
+        while i > 0:
+            prod *= dims[i]
+            total += prod * indices[i-1]
+            i -= 1
+        return total
+    
+    def mem_load_dn(self, mem, dims, indices, reg, groupname):
+        """Inserts wiring into `self` to perform `reg := mem[i1][i2]...[in]`,
+        where `mem` is a seq_dn memory or a comb_mem_d1 memory
```

Comment:
Here, also explain the use-case of how `indices` has to be used. And consider the assertion that Rachit suggests below!